service/listeners.go

```diff
@@ -128,42 +128,33 @@ type readRequest struct {
 
 type virtualPacketConn struct {
 	net.PacketConn
-	mu          sync.Mutex // Mutex to protect access to the channels
 	readCh      chan readRequest
+	closeCh     chan struct{}
 	onCloseFunc OnCloseFunc
 }
 
-func (pc *virtualPacketConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+func (pc *virtualPacketConn) ReadFrom(p []byte) (int, net.Addr, error) {
 	respCh := make(chan struct {
 		n    int
 		addr net.Addr
 		err  error
 	}, 1)
 
-	pc.mu.Lock()
-	if pc.readCh == nil {
-		pc.mu.Unlock()
-		return 0, nil, net.ErrClosed
-	}
-	pc.readCh <- readRequest{
+	select {
+	case pc.readCh <- readRequest{
```

Comment:
Let's check if the read is ok here just to be safe, in case the channel was closed.

Reply:
I don't think you can do that here. As I understand it, sending to a channel in a select statement will only proceed if the channel is ready. If the channel is closed, this won't be selected.

Reply:
A read from a closed channel always succeed immediately, with the zero value. You need to check the `ok` flag to detect that.

(on the other hand, `nil` channels always block)